### DISCUSSION TITLE
Realm key(s) uses - Allowing multivalue fields

### DISCUSSION THREAD

**Discussion by laskasn** [keycloak/keycloak#8523] [Ideas]
Base [PR](https://github.com/keycloak/keycloak/pull/8539) for key uses
Additional [ PR](https://github.com/keycloak/keycloak/pull/8684) for key protocol uses

### Motivation:

Currently, Keycloak allows an administrator to define multiple keys of different type and priority, along with an intended use (for signing or encryption). Moreover, the current design allows the keys to be used exclusively for one use, either signing (sig) or encryption (enc). So, in case an admin has a RSA-256 key marked for sig, but wishes to use the same key for enc as well, he/she needs to duplicate this key value, which is not efficient.

Furthermore, it is currently not possible to configure different keys per authentication protocol, e.g. SAML, OpenID Connect (OIDC), kerberos, etc. For instance, SAML and OIDC certificates typically have different rotation requirements, i.e. in SAML every 10 years while in OIDC every 6 six months. So, we would like to propose an extension to allow the realm administrator to use different keys for each protocol.

### Proposal:

We plan to contribute our changes in two pull requests:

- a small one about the enc/sig uses of a key, 
- an extension of that, which allows determining the protocol use for a key (saml, oidc, etc). 

**Pull request 1:**

So, for the first one, we change the exclusiveness of each key’s use for either “enc” or “sig”, and -instead- assign a list of uses for each key i.e. [“enc”, “sig”].

So, if an admin needs to use an existing key for more than one use, he/she just needs to add the appropriate use tag for that key, instead of replicating the whole key which requires more effort and is error-prone.
To achieve that, few changes are required: 

- we switch the “use” into a multivalue field (it is already a multivalue field, but with always a single value)
- we slightly modify the keys fetching functions, to adapt to this multi-valued “use” change (change equals() with contains() )
- finally, we change the UI to be able to handle multiple values for the “use” field (also allow adding array values or removing). That’s supported out-of the box, by changing the ProviderConfigProperty type for the UI field.

Here’s a screenshot of the modified field (UI).
![this](https://user-images.githubusercontent.com/11722418/135611387-8e380626-ce78-45a4-80e2-eaca22cf01b2.png)

**Pull request 2:**

In the second pull request, we add support for different key uses by (auth) protocol. We plan to add an additional field (i.e “protocol-use”), having the same tagging design as the sig/enc use, but assigning a list of “permitted to use” with protocols.

Of course, as with the sig/enc modification, we 

- have to change the “fetching by” functions to be also parametrized by the authentication protocol 
- have to change the key editing UI by adding the extra field

Protocol values should exist as an enum in the java code. For instance, the protocol enumerations could be {“oidc”, “saml”, “kerberos”, … , “other”} 
For backwards compatibility in case of upgrades, if this new field is missing, it’ll be considered as “all”, but of course, the admin can easily edit it by adding the desired value(s).

### Additional details and examples:

IMPORTANT: If keys are also linked to a protocol type (oidc, saml, kerberos, etc), without having multi-value fields, then the combinations escalate quickly, since an admin should need to define M*N keys, with M the # of uses {sig,enc} and N the # of protocols {oidc, saml, kerberos, ldap, ...} 

So, instead of having a list of RS256 keys (for sake of simplicity, we omit HMAC, AES and ECDSA)

![1](https://user-images.githubusercontent.com/11722418/135611654-b46a712b-e0ef-4684-947e-408c9c7f6304.png)

by using multi-value fields, an admin will be able to define the equivalent: 

![2](https://user-images.githubusercontent.com/11722418/135611742-5681bff4-e161-4d40-a47f-6f1a2dade6a2.png)

being able to add/remove tags from the “use” and “protocol” lists.

Currently, upon initialization (or realm creation), keycloak checks if the minimum required keys for keycloak to work exist in the realm, and if not, generates them. 

With the proposed tagging scheme, keycloak will only need to check for one key of each type rsa,hsa,etc and assign the minimum tags required to work.

So, by default it would generate something like this for each new realm (including the “master”):
![3](https://user-images.githubusercontent.com/11722418/135611805-082da51e-90fa-4565-80de-dda92d069917.png)

If an admin would like to change the RSA key for -let’s say- ldap, he should edit the existing row by removing the ldap tag from the rsa protocol list
![4](https://user-images.githubusercontent.com/11722418/135611867-e04246d7-a515-4892-896b-af03788b82a8.png)
and create a new entry with 
![5](https://user-images.githubusercontent.com/11722418/135611922-385f88a1-f1fc-47d8-964b-c16023a342db.png)

**Comment by stianst** [comment]
Keys was initially designed to have a single use to prevent any potential vulnerabilities around multiple uses of the same keys. However, I've been trying to think of a potential way keys that are used both of sign and enc could be a vulnerability, but haven't been able to come up with any issues there. 

If memory serves me right this came a bit out of the RSA/HMAC vulnerability in JWKs, but that's when a key is used for different algorithms, rather than different uses.

There's still something that bothers me slightly about this, but can't provide any evidence, other than JWK specs do not permit multiple values for use.

**Reply by laskasn** [reply]
I'm not aware of any vulnerabilities  around multiple uses of the same keys either. However, if the realm administrator chooses to use the same key for both signing and encryption, it would be easier to define the key once and specify the different uses instead of having to define the same key multiple times, one per use. 

The spec, defines: 
4.2.  "use" (Public Key Use) Parameter
   The "use" (public key use) parameter identifies the intended use of
   the public key.  The "use" parameter is employed to indicate whether
   a public key is used for encrypting data or verifying the signature
   on data.
   Values defined by this specification are:
   o  "sig" (signature)
   o  "enc" (encryption)
   Other values MAY be used.  The "use" value is a case-sensitive
   string.  **Use of the "use" member is OPTIONAL, unless the application
   requires its presence.**
   When a key is used to wrap another key and a public key use
   designation for the first key is desired, the "enc" (encryption) key
   use value is used, since key wrapping is a kind of encryption.  The
   "enc" value is also to be used for public keys used for key agreement
   operations.
   Additional "use" (public key use) values can be registered in the
   IANA "JSON Web Key Use" registry established by Section 8.2.
   Registering any extension values used is highly recommended when this
   specification is used in open environments, in which multiple
   organizations need to have a common understanding of any extensions
   used.  However, unregistered extension values can be used in closed
   environments, in which the producing and consuming organization will
   always be the same.
https://www.ietf.org/rfc/rfc7517.txt


By default, a realm gets intialized with 2 distinct RSA keys: one for "sig" and one for "enc".
If we add another one (RSA key), this time with both "enc" and "sig" uses - as seen in the following screenshot,

![keys_config](https://user-images.githubusercontent.com/11722418/135872581-9d9274a4-c30b-413b-bb8f-6e05e63b4f1b.png)

then the result of the `https://<KEYCLOAK_HOST>/auth/realms/<REALM>/protocol/openid-connect/certs` is the following:

![JWKS](https://user-images.githubusercontent.com/11722418/135872879-e64960c5-f617-471e-b8c3-5db4042a388d.png)


Just for comparison, with the current vanilla keycloak v15.0.2, if we add another rsa key for "sig" and also another one for "enc", and end up with this:


![vanilla-keys-config](https://user-images.githubusercontent.com/11722418/135875768-a8a732ca-e91c-4002-9823-8bb77e89698a.png)

then the same call to the jwks endpoint, gives you the following output:

 
![vanilla-jwks](https://user-images.githubusercontent.com/11722418/135875852-ae3ad194-8db4-40c6-8a8b-85f3788942ad.png)


We submit a pull request with this change, to share this with the community

**Reply by laskasn** [reply]
The pull request for the **pr proposal 1** can be found [here](https://github.com/keycloak/keycloak/pull/8539)

**Reply by abstractj** [reply]
> Keys was initially designed to have a single use to prevent any potential vulnerabilities around multiple uses of the same keys. However, I've been trying to think of a potential way keys that are used both of sign and enc could be a vulnerability, but haven't been able to come up with any issues there.
> 
> If memory serves me right this came a bit out of the RSA/HMAC vulnerability in JWKs, but that's when a key is used for different algorithms, rather than different uses.
> 
> There's still something that bothers me slightly about this, but can't provide any evidence, other than JWK specs do not permit multiple values for use.

@stianst your concern is valid and key reuse is not a recommended practice. For reference, I suggest the reading of [The Dangers of Key Reuse: Practical Attacks on IPsec IKE](https://www.usenix.org/system/files/conference/usenixsecurity18/sec18-felsch.pdf). Even though we don't have any practical attacks against Keycloak, that's something to keep in mind.

**Reply by SecKatie** [reply]
I think the largest issue might come up in the future.

If a key can have multiple uses what will prevent a key from being used for multiple algorithms in the future when we have to upgrade Keycloak to utilize a newer algorithm to prevent a, currently unknown, attack on RS256? Using a key for enc and sig (in the case of RSA) is not a problem but it might be a problem for future algos.

**Comment by laskasn** [comment]
I’ve read the paper. Indeed, there is a chance that someone might perform a Bleichenbacher attack on the IKE, but the flaw they describe is actually in protocol(s) where there is a need to perform a key exchange. Is this really affecting the use of the same key in encryption and signing of let’s say SAML assertions? That’s something i'm missing. 

Checking the bibliography about that, i stumbled upon [this](https://crypto.stanford.edu/RealWorldCrypto/slides/kenny.pdf) nice presentation which includes some relevant research (published papers) on using the same RSA keys for encryption and signatures.

However, what we propose, is not going to affect the existing default setup: **each realm is always initialized with a different rsa key for encryption and signing**. Whether extending the use of one of them to both uses, is a decision of the keycloak admin (should be on purpose changed from the admin UI). This can be already done in the existing vanilla keycloak by just duplicating the key for the other use.

I personally believe that more problems arise from choosing to use legacy cryptosystems, rather than from reusing the same key in enc and sig. For instance, see [here](https://www.nds.ruhr-uni-bochum.de/media/nds/veroeffentlichungen/2012/12/19/XMLencBleichenbacher.pdf)

**Comment by stianst** [comment]
Just to confirm.. We all agree that unique keys per-use is the ideal? However, in practice that's not always practical, and end of the day may not make much difference from a security perspective?

**Reply by laskasn** [reply]
I completely agree with your statement!

**Reply by stianst** [reply]
How could we then make it a bit more obvious that the recommendation is single-use? With the PR right now, and the multi-input field, isn't it somewhat guiding users to multi-use being recommended?

**Reply by laskasn** [reply]
By default, it will have different keys for each use. But, yes, you are right, we should add something to make it clear that it's highly advisable to use different keys for each use.
I believe we can add it in the existing tooltip or if you think that this is not enough, we could also add a confirmation popup like the ones keycloak uses  i.e. for delete, which will appear whenever user tries to add both enc/sig uses.
However, i believe that the current default design is to have a different key per use (we don't change anything regarding that), so, i guess in case the admin wants to mess up with the keys, he/she should at least be aware of what he/she is doing.

**Reply by stianst** [reply]
The confirmation popup makes it more obvious for sure, but perhaps a bit too intrusive. I'm okay with just making it clear in the tooltip.

**Reply by abstractj** [reply]
+1 and makes sense to add that as a tooltip that's not something recommended.

**Reply by laskasn** [reply]
Added to the tooltip (see file [Attributes.java](https://github.com/keycloak/keycloak/pull/8539/files) ) the following text: 
`Although not prohibited, it's strongly advised to configure a single use for each key.`
So it now reads: 
`Whether the key should be used for signing or encryption. Although not prohibited, it's strongly advised to configure a single use for each key.`
Is this sufficient? Should we elaborate more?

**Reply by jbman** [reply]
Language hint: rather than _not prohibited_ it should read _not forbidden_ (Prohibited is by law)

**Reply by laskasn** [reply]
Thanks for the tip jbman! I'll update the text as suggested!
